### PR TITLE
Set ExpansionFactor to 0.06 when it's equals to 0 in the private dict of CFF fonts

### LINKED ISSUE #14759: Chinese font not displayed
Attach (recommended) or Link to PDF file here:
[GM／T 0016-2012 智能密码钥匙密码应用接口规范(1).pdf](https://github.com/mozilla/pdf.js/files/8441278/GM.T.0016-2012.1.pdf)

Configuration:
- Web browser and its version: Microsoft Edge 99.0.1150.39
- Operating system and its version: Windows 10
- PDF.js version: v2.13.216
- Is a browser extension: No

Steps to reproduce the problem:
1. Open the PDF with pdf.js.
2. Found the text of this PDF disappeared.

What is the expected behavior? (add screenshot)
Open with browser built-in viewer
image

What went wrong? (add screenshot)
Open with pdf.js online demo https://mozilla.github.io/pdf.js/web/viewer.html
image
image

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

### LINKED ISSUE #15289: Showing squares with hex values instead of text in some PDFs
Attach (recommended) or Link to PDF file here:
* Any O'Reilly book preview PDF at [Humble Bundle](https://www.humblebundle.com) seems to have the problem. You can find a current bundle [here](https://www.humblebundle.com/books/popular-programming-languages-oreilly-books) with many direct links. Please note the bundles only last for a limited time, and direct links to preview PDFs have a "ttl" query string parameter so I suppose they expire. I'm unwilling to upload an example here since I don't own the content. It would be ideal if a pdf.js dev could download test cases themselves while the bundle is still active.

Configuration:
- Web browser and its version: Firefox 103.0.1 (64-bit).
- Operating system and its version: Windows 10 Pro 10.0.19044.
- PDF.js version: 2.15.129 according to the developer tools console.
- Is a browser extension: No.

Steps to reproduce the problem:
1. Open one of the O'Reilly book preview PDFs on Humble Bundle and see the problem. It also happens if I save the preview PDF locally and then open it. Others have reported the problem on Reddit [here](https://old.reddit.com/r/humblebundles/comments/w4n0ws/broken_pdf_preview_on_firefox/). It only started recently, in the past month or so. If I copy and paste text from the broken preview PDF into Notepad, the text looks fine.
2. The problem doesn't happen if I:
* open a complete O'Reilly non-preview PDF that is local on my system
* open an O'Reilly preview PDF in Firefox on Linux
* open an O'Reilly preview PDF in Chrome on Windows
* open a Humble Bundle preview PDF from another publisher (Packt)

What is the expected behavior? (add screenshot)
* PDF is readable.

What went wrong? (add screenshot)
* PDF is unreadable because all the text is replaced by squares with hex values. On some but not all broken PDFs I see errors like this in the console:

Warning: Failed to load font 'g_d0_f3': 'SyntaxError: An invalid or illegal string was specified'. pdf.js:446:13
downloadable font: CFF : Failed to parse Global Subrs INDEX (font-family: "g_d0_f3" style:normal weight:400 stretch:100 src index:0) source: (invalid URI)
downloadable font: CFF : Failed to parse table (font-family: "g_d0_f3" style:normal weight:400 stretch:100 src index:0) source: (invalid URI)
downloadable font: rejected by sanitizer (font-family: "g_d0_f3" style:normal weight:400 stretch:100 src index:0) source: (invalid URI)
downloadable font: font load failed (font-family: "g_d0_f3" style:normal weight:400 stretch:100 src index:0) source: (invalid URI) 

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):
* N/A.

### LINKED ISSUE #15902: Blank PDF but the text layer isn't empty
Attach (recommended) or Link to PDF file here:
[code-debug-gs.pdf](https://github.com/mozilla/pdf.js/files/10366793/code-debug-gs.pdf)

The pdf comes from: https://gitlab.freedesktop.org/poppler/poppler/-/issues/1327

Configuration:
- Web browser and its version: Firefox nightly
- Operating system and its version: Windows 11

The pdf is blank when some text is displayed in Chrome, Acrobat and in Okular.

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

File: test/test_manifest.json [hunk 1]
```diff
@@ -7292,5 +7292,12 @@
       "rounds": 1,
       "link": true,
       "type": "eq"
+   },
+   {
+      "id": "issue15289",
+      "file": "pdfs/issue15289.pdf",
+      "md5": "f3d20aee398e052ac3dce7cc030f50de",
+      "rounds": 1,
+      "type": "eq"
```
Comment (Snuffleupagus) [inline test/test_manifest.json:7301]: It should suffice to test only one page, right?
```suggestion
   {
      "id": "issue15289",
      "file": "pdfs/issue15289.pdf",
      "md5": "f3d20aee398e052ac3dce7cc030f50de",
      "rounds": 1,
      "lastPage": 1,
      "type": "eq"
```

**Review by Snuffleupagus — APPROVED** [review]
That's a fascinating find, how did you figure it out?

r=me, with the comment addressed and passing tests; thank you!

**Comment by calixteman** [comment]
> That's a fascinating find, how did you figure it out?
> 

I compared the original font I extracted from the pdf and the one we generate.
I ran `tx` on both fonts and noticed some differences in the charsets, so I manually fix the charset to have exactly the same as in the original font but nothing.
I ran `fontlint` on the generated font and noticed few errors about `BlueValues`, I fixed them but nothing. I was pretty sure that something was wrong in the top dict.
Finally I had the wonderful idea to just remove the private dict and magically it worked, then I just tried to remove the properties one by one and finally the last one (as usual) was the culprit.

File: test/test_manifest.json [hunk 2]
```diff
@@ -7292,5 +7292,13 @@
...
       "link": true,
       "type": "eq"
+   },
+   {
+      "id": "issue15289",
+      "file": "pdfs/issue15289.pdf",
+      "md5": "f3d20aee398e052ac3dce7cc030f50de",
+      "rounds": 1,
+      "lastPage": 1,
+      "type": "eq"
```
Comment (Snuffleupagus) [inline test/test_manifest.json:7302]: There's already a test-case for this document, hence you'll need to remove all the changes to the `test/`-folder in this patch.

**Comment by calixteman** [comment]
@Snuffleupagus instead of removing the property I set it to 0.06 to follow Jonathan's advice: https://github.com/mozilla/pdf.js/issues/15289#issuecomment-1374484221
Are you ok with this change ?

**Comment by Snuffleupagus** [comment]
> Are you ok with this change ?

Yes, since I do agree that it sounds like an even safer approach here.

Given the amount of duplicates, we might want to uplift this to previous Firefox versions?

---

As a follow-up to this PR, to prevent future bugs, I do wonder if we want to add similar code when parsing "regular" Type1 fonts as well or if should we perhaps wait until an actual bug is reported?
https://github.com/mozilla/pdf.js/blob/d6984f93954a8ac9e92cc9f156f7e909d3e332de/src/core/type1_parser.js#L657

**Pull request merged**